src/EditorFeatures/Core/Implementation/GoToImplementation/AbstractGoToImplementationService.cs

```diff
@@ -0,0 +1,75 @@
+namespace Microsoft.CodeAnalysis.Editor.Implementation.GoToImplementation
+{
+    abstract class AbstractGoToImplementationService : IGoToImplementationService
+    {
+        private readonly IEnumerable<Lazy<INavigableItemsPresenter>> _navigableItemPresenters;
+
+        public AbstractGoToImplementationService(IEnumerable<Lazy<INavigableItemsPresenter>> navigableItemPresenters)
+        {
+            _navigableItemPresenters = navigableItemPresenters;
+        }
+
+        public bool TryGoToImplementation(Document document, int position, CancellationToken cancellationToken, out string message)
+        {
+            var symbol = SymbolFinder.FindSymbolAtPositionAsync(document, position, cancellationToken).WaitAndGetResult(cancellationToken);
+
+            if (symbol != null)
+            {
+                var solution = document.Project.Solution;
+                var implementations =
+                    SymbolFinder.FindImplementationsAsync(
+                        symbol,
+                        solution,
+                        cancellationToken: cancellationToken).WaitAndGetResult(cancellationToken).ToList();
+
+                if (implementations.Count == 0)
+                {
+                    message = EditorFeaturesResources.SymbolHasNoImplementations;
+                    return false;
+                }
+                else if (implementations.Count == 1)
+                {
+                    GoToDefinition.GoToDefinitionHelpers.TryGoToDefinition(implementations.Single(), document.Project, _navigableItemPresenters, cancellationToken);
+                    message = null;
+                    return true;
+                }
+                else
+                {
+                    // We have multiple symbols, so we'll build a list of all preferred locations for all the symbols
+                    var navigableItems = implementations.SelectMany(
+                        implementation => CreateItemsForImplementation(implementation, solution));
+
+                    var presenter = _navigableItemPresenters.First();
+                    presenter.Value.DisplayResult(NavigableItemFactory.GetSymbolDisplayString(document.Project, symbol), navigableItems);
+                }
+
+                message = null;
+                return true;
+            }
+            else
+            {
+                message = EditorFeaturesResources.CannotNavigateToTheSymbol;
+                return false;
+            }
+        }
+
+        private static IEnumerable<INavigableItem> CreateItemsForImplementation(ISymbol implementation, Solution solution)
+        {
+            var symbolDisplayService = solution.Workspace.Services.GetLanguageServices(implementation.Language).GetRequiredService<ISymbolDisplayService>();
+
+            return NavigableItemFactory.GetItemsFromPreferredSourceLocations(solution, implementation,
+                                        displayString: symbolDisplayService.ToDisplayString(implementation));
```

Comment:
nit: the alignment here looks weird to me. I'd just wrap all of the arguments and have a single indent.